darts/utils/likelihood_models.py

```diff
@@ -1,32 +1,150 @@
+
+Some distributions (such as ``GaussianLikelihood``, and ``PoissonLikelihood``) are univariate,
+in which case they are applied to model each component of multivariate series independently.
+Some other distributions (such as ``DirichletLikelihood``) are multivariate,
+in which case they will model all components of multivariate time series jointly.
+
+Univariate likelihoods accept either scalar or array-like values for the optional prior parameters.
+If a scalar is provided, it is used as a prior for all components of the series. If an array-like is provided,
+the i-th value will be used as a prior for the i-th component of the series. Multivariate likelihoods
+require array-like objects when specifying priors.
+
+The target series used for training must always lies within the distribution's support, otherwise
+errors will be raised during training. You can refer to the individual likelihoods' documentation
+to see what is the support. Similarly, the prior parameters also have to lie in some pre-defined domains.
 """
 
+# TODO: Table on README listing distribution, possible priors and wiki article
+from darts.utils.utils import raise_if_not
+
 from abc import ABC, abstractmethod
+from typing import Optional, Tuple
+import collections
+
+import numpy as np
 import torch
 import torch.nn as nn
+from torch.distributions.kl import kl_divergence
+from torch.distributions import (Normal as _Normal,
+                                 Poisson as _Poisson,
+                                 NegativeBinomial as _NegativeBinomial,
+                                 Bernoulli as _Bernoulli,
+                                 Gamma as _Gamma,
+                                 Gumbel as _Gumbel,
+                                 Laplace as _Laplace,
+                                 Beta as _Beta,
+                                 Exponential as _Exponential,
+                                 MultivariateNormal as _MultivariateNormal,
+                                 Dirichlet as _Dirichlet,
+                                 Geometric as _Geometric,
+                                 Cauchy as _Cauchy,
+                                 ContinuousBernoulli as _ContinuousBernoulli,
+                                 HalfNormal as _HalfNormal,
+                                 LogNormal as _LogNormal,
+                                 Weibull as _Weibull
+                                 )
 
+MIN_CAUCHY_GAMMA_SAMPLING = 1e-100
 
-class LikelihoodModel(ABC):
+# Some utils for checking parameters' domains
+def _check(param, predicate, param_name, condition_str):
+    if param is None: return
+    if isinstance(param, (collections.Sequence, np.ndarray)):
+        raise_if_not(all(predicate(p) for p in param),
+                     'All provided parameters {} must be {}.'.format(param_name, condition_str))
+    else:
+        raise_if_not(0 < param, 'The parameter {} must be {}.'.format(param_name, condition_str))
```

Comment:
```suggestion
        raise_if_not(predicate(param), 'The parameter {} must be {}.'.format(param_name, condition_str))
```
..unless I'm not understanding something

Reply:
Good spot!